platforms/core-execution/persistent-cache/src/main/java/org/gradle/cache/internal/CacheVersionMapping.java

```diff
@@ -65,6 +65,10 @@ public Builder incrementedIn(String minGradleVersion) {
             return changedTo(versions.get(versions.lastKey()) + 1, minGradleVersion);
         }
 
+        public Builder retiredIn(String gradleVersion) {
+            return changedTo(Integer.MAX_VALUE, gradleVersion);
```

Comment:
❓ Is this retiredIn required?

Asking this since then this `Integer.MAX_VALUE` is added to versionMapping and I am not sure what are consequences of that. So I wonder if it's better to not add this version there.

Reply:
The `CacheVersionMapping` is used to determine if a particular cache directory is used by a version of Gradle. The `UnusedVersionsCacheCleanup` looks at the currently used Gradle versions and removes any cache directories that are not used by any of them.

I added the `retiredIn` method to make it clear that there is no valid `transforms-X` directory for that gradle version (and newer). Setting the version to `Integer.MAX_VALUE` was a simple implementation.
I guess it would be better to implement `retiredIn` such that [these methods](https://github.com/gradle/gradle/blob/497dffa06e9a0769c5e0e2ba866d0fbd88f7a1d5/platforms/core-execution/persistent-cache/src/main/java/org/gradle/cache/internal/CacheVersionMapping.java#L44-L51) return `null` instead of `Integer.MAX_VALUE`. But the overall impact will be the same.

Reply:
I tried a few other implementations, but they all added complexity without really adding any value. If `CacheVersionMapping` returns `null` for a "retired" cache layout, then all of the calling code needs to deal with null.
I've left this as-is: eventually we may be able to remove the `TRANSFORMS` mapping altogether. But for now it is required so that new Gradle versions are able to clean up the `transforms-X` directories left behind by older versions.

Reply:
Ok, sounds good to me, thanks for checking other options